source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,118 @@
+Functions as Data
+=================
+
+Regardless of programming language, one of the first things anyone learns to do when programming a computer is to write a function (also known as a "method" or a "subroutine").  Functions are a fundamental part of organized code - writing functions lets us avoid duplicating the same piece of code over and over again.  Instead of writing duplicated sections of code, we call a single function that contains the code we want to execute from multiple places (provided we named the function well, the function name is also easier to read than the code itself!).  If the section of code needs some additional information about its surrounding context to run, we pass those to the function as "parameters", and iff it needs to yield something back to the rest of the code once it finishes, we call that a "return value" (together, the parameters and return value are called the function's "signature");
+
+Sometimes, we need to pass functions from one part of the code to another part of the code.  This might seem like a strange concept, if we're used to thinking of functions as part of a class definition rather than objects in their own right.  But at a basic level, functions are just data - in the same way we can store an ``integer`` or a ``double`` as a variable and pass it around our program, we can do the same thing with a function.  A variable whose value is a function is called a "functional interface" in Java, and a "function pointer" or "functor" in C++.
+
+Why Would We Want to Treat Functions as Data?
+---------------------------------------------
+
+Typically, code that calls a function is coupled to (depends on) the definition of the function. While this occurs all the time, it becomes problematic when the code *calling* the function (for example, WPILib) is developed independently and without direct knowledge of the code that *defines* the function (for example, code from an FRC team). Sometimes we solve this challenge through the use of class interfaces, which define collections of data and functions that are meant to be used together.  However, often we really only have a dependency on a *single function*, rather than on an *entire class*.
+
```

Comment:
I feel like this why is a little bit too dry and something a student might have a hard time following.  How about injecting something like this:
For example, WPILib provides great facilities for handling joysticks.  It handles a lot of  the problems of tracking button state and watching for a button press and release.  But then we  want to write code that achieves something like:  'When this button is pressed, move the arm up'.  So if we have a function that 'moves the arm up' it is very powerful if we can pass that function in to the WPILib joystick methods.

Reply:
I like that a lot, and will incorporate it.

Reply:
Generally, I felt like describing what a function is (the first paragraph) was something someone should understand if they are reading this page, so you could do without that whole paragraph.  Starting just with the second paragraph gets more to the point of functions as data.  It may be appropriate to make the first mention of function there have a mouse-over definition, to make clear that methods and functions are the same.

Reply:
I disagree.  I liked that paragraph.  The whole point of this is that WPILib is using Lambda expressions in the very first example code a student programmer might use.   If anything, I feel that this article should be studied and made sure it is comprehensible to someone with zero previous programming experience.